crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,725 @@
+impl CommutationChecker {
+    #[pyo3(signature = (standard_gate_commutations=None, cache_max_entries=1_000_000, gates=None))]
+    #[new]
+    fn py_new(
+        standard_gate_commutations: Option<Bound<PyAny>>, // Send a bound here
+        cache_max_entries: usize,
+        gates: Option<HashSet<String>>,
+    ) -> Self {
+        // Initialize sets before they are used in the commutation checker
+        Lazy::force(&SKIPPED_NAMES);
+        Lazy::force(&NO_CACHE_NAMES);
+        Lazy::force(&SUPPORTED_OP);
+
+        CommutationChecker {
+            library: CommutationLibrary::new(standard_gate_commutations),
+            cache: HashMap::with_capacity(cache_max_entries),
+            cache_max_entries,
+            current_cache_entries: 0,
+            _cache_miss: 0,
+            _cache_hit: 0,
+            gates,
+        }
+    }
+
+    #[pyo3(signature=(op1, op2, max_num_qubits=3))]
+    #[allow(clippy::too_many_arguments)]
+    fn commute_nodes(
+        &mut self,
+        py: Python,
+        op1: &DAGOpNode,
+        op2: &DAGOpNode,
+        max_num_qubits: u32,
+    ) -> PyResult<bool> {
+        let mut bq: BitData<Qubit> = BitData::new(py, "qubits".to_string());
+        let op1_bound_qubits = op1.instruction.qubits.bind(py);
+        let op2_bound_qubits = op2.instruction.qubits.bind(py);
+        op1_bound_qubits
+            .iter()
+            .for_each(|q| bq.add(py, &q, false).unwrap());
+        op2_bound_qubits
+            .iter()
+            .for_each(|q| bq.add(py, &q, false).unwrap());
+        let qargs1 = op1_bound_qubits
+            .iter()
+            .map(|q| bq.find(&q).unwrap().0)
+            .collect::<Vec<_>>();
+        let qargs2 = op2_bound_qubits
+            .iter()
+            .map(|q| bq.find(&q).unwrap().0)
+            .collect::<Vec<_>>();
+
+        let mut bc: BitData<Clbit> = BitData::new(py, "clbits".to_string());
+        let op1_bound_clbit = op1.instruction.clbits.bind(py);
+        let op2_bound_clbit = op2.instruction.clbits.bind(py);
+
+        op1_bound_clbit
+            .iter()
+            .for_each(|c| bc.add(py, &c, false).unwrap());
+
+        op2_bound_clbit
+            .iter()
+            .for_each(|c| bc.add(py, &c, false).unwrap());
+
+        let cargs1 = op1_bound_clbit
+            .iter()
+            .map(|c| bc.find(&c).unwrap().0)
+            .collect::<Vec<_>>();
+        let cargs2 = op2_bound_clbit
+            .iter()
+            .map(|c| bc.find(&c).unwrap().0)
+            .collect::<Vec<_>>();
+
+        Ok(self.commute_inner(
+            py,
+            &op1.instruction,
+            &qargs1,
+            &cargs1,
+            &op2.instruction,
+            &qargs2,
+            &cargs2,
+            max_num_qubits,
+        ))
+    }
+
+    #[pyo3(signature=(op1, qargs1, cargs1, op2, qargs2, cargs2, max_num_qubits=3))]
+    #[allow(clippy::too_many_arguments)]
+    fn commute(
+        &mut self,
+        py: Python,
+        op1: &Bound<PyAny>,
```

Comment:
You can change this to:

```suggestion
        op1: &Bound<OperationFromPython>,
```

and then skip the `CircuitInstruction` creation altogether (if you follow my suggestion for the function signature change). `OperationFromPython` will give you all the data you need to check commutation directly and implements `FromPyObject` so you don't need any intermediate python object creation.

Reply:
The compiler doesn't seem to like this idea, it's telling me that
```
the trait bound `&pyo3::Bound<'_, OperationFromPython>: PyFunctionArgument<'_, '_>` is not satisfied
the trait `PyFunctionArgument<'a, 'py>` is implemented for `&'a pyo3::Bound<'py, T>
```
maybe just implementing `FromPyObject` is not enough?

Reply:
Ah my bad sorry, yeah it needs be `&OperationFromPython` you don't need the `Bound` since it's not a python type.